loader/db.go

```diff
@@ -34,213 +36,151 @@ import (
-	return conn.executeSQLCustomRetry(ctx, sqls, enableRetry, isRetryableError)
-}
 
-func (conn *Conn) executeDDL(ctx *tcontext.Context, sqls []string, enableRetry bool) error {
-	return conn.executeSQLCustomRetry(ctx, sqls, enableRetry, isDDLRetryableError)
+	return conn.executeSQLCustomRetry(ctx, sqls, retry.IsLoaderDDLRetryableError)
 }
 
-func (conn *Conn) executeSQLCustomRetry(ctx *tcontext.Context, sqls []string, enableRetry bool, isRetryableFn func(err error) bool) error {
-	if len(sqls) == 0 {
+func (conn *Conn) executeSQL(ctx *tcontext.Context, queries []string, args ...[]interface{}) error {
+	if len(queries) == 0 {
 		return nil
 	}
 
-	if conn == nil || conn.db == nil {
-		return terror.ErrDBUnExpect.Generate("database connection not valid")
-	}
-
-	var err error
-
-	retryCount := 1
-	if enableRetry {
-		retryCount = maxRetryCount
-	}
-
-	for i := 0; i < retryCount; i++ {
-		if i > 0 {
-			ctx.L().Debug("execute statement", zap.Int("retry", i), zap.String("sqls", fmt.Sprintf("%-.200v", sqls)))
-			time.Sleep(2 * time.Duration(i) * time.Second)
-		}
-
-		startTime := time.Now()
-		err = executeSQLImp(ctx, conn.db, sqls)
-
-		failpoint.Inject("LoadExecCreateTableFailed", func(val failpoint.Value) {
-			items := strings.Split(val.(string), ",")
-			if len(items) != 2 {
-				ctx.L().Fatal("failpoint LoadExecCreateTableFailed's value is invalid", zap.String("val", val.(string)))
-			}
-
-			errCode, err1 := strconv.ParseUint(items[0], 10, 16)
-			errNum, err2 := strconv.ParseInt(items[1], 10, 16)
-			if err1 != nil || err2 != nil {
-				ctx.L().Fatal("failpoint LoadExecCreateTableFailed's value is invalid", zap.String("val", val.(string)), zap.Strings("items", items), zap.Error(err1), zap.Error(err2))
-			}
-
-			if i < int(errNum) && len(sqls) == 1 && strings.Contains(sqls[0], "CREATE TABLE") {
-				err = tmysql.NewErr(uint16(errCode))
-				ctx.L().Warn("executeSQLCustomRetry failed", zap.String("failpoint", "LoadExecCreateTableFailed"), zap.Error(err))
-			}
-		})
-
-		if err != nil {
-			tidbExecutionErrorCounter.WithLabelValues(conn.cfg.Name).Inc()
-			if isRetryableFn(err) {
-				continue
-			}
-			return terror.DBErrorAdapt(err, terror.ErrDBExecuteFailed, strings.Join(sqls, ";"))
-		}
-
-		// update metrics
-		cost := time.Since(startTime)
-		txnHistogram.WithLabelValues(conn.cfg.Name).Observe(cost.Seconds())
-		if cost > 1 {
-			ctx.L().Warn("transaction execute successfully", zap.Duration("cost time", cost))
+	sqls := make([]baseconn.SQL, 0, len(queries))
+	for i, query := range queries {
+		if i >= len(args) {
+			sqls = append(sqls, baseconn.SQL{Query: query, Args: []interface{}{}})
+		} else {
+			sqls = append(sqls, baseconn.SQL{Query: query, Args: args[i]})
 		}
-
-		return nil
 	}
 
-	return terror.DBErrorAdapt(err, terror.ErrDBExecuteFailed, strings.Join(sqls, ";"))
+	return conn.executeSQLCustomRetry(ctx, sqls, retry.IsLoaderRetryableError)
```

Comment:
only this line is different from `executeDDL`? can we refactor it to decrease duplicate code?

Reply:
done in https://github.com/pingcap/dm/pull/236/commits/59dacaa6bceb5fd19f6360808d5875fdb48a01c5#diff-5f5bb90e4d87dad0cb25c377bb10679bR63